### PR TITLE
Code Coverage: Report Service #5

### PR DESCRIPTION
**Pull request by jayfyle**
## Code coverage almost 100%

Report_Coverage_98

### CODE REVIEW

File: src/app/core/mock-data/api-reports.data.ts [hunk 1]
```diff
@@ -547,8 +547,8 @@ export const expectedReports: ApiV2Response<ExtendedReport> = {
   url: '/v2/reports',
 };
 
-export const apiAllReportsRes: ApiV2Response<ExtendedReport> = {
-  count: 2,
+export const apiAllReportsRes1: ApiV2Response<ExtendedReport> = {
+  count: 3,
```
Comment (Julias0) [inline src/app/core/mock-data/api-reports.data.ts:551]: Why is this 3? There are 2 items in data and the offset is 0, limit 2 so this cant happen right?

File: src/app/core/mock-data/api-reports.data.ts [hunk 2]
```diff
@@ -662,6 +662,69 @@ export const apiAllReportsRes: ApiV2Response<ExtendedReport> = {
   url: '/v2/reports',
 };
 
+export const apiAllReportsRes2: ApiV2Response<ExtendedReport> = {
+  count: 3,
```
Comment (Julias0) [inline src/app/core/mock-data/api-reports.data.ts:666]: Why is this with count 3 when there is just 1 line item?

File: src/app/core/models/report-unflattened.model.ts [hunk 3]
```diff
@@ -6,12 +7,14 @@ export interface UnflattenedReport {
   ou: OrgUser;
   rp: {
     amount: number;
-    approvals?: {
-      [id: string]: {
-        rank: number;
-        state: string;
-      };
-    };
+    approvals?:
```
Comment (Julias0) [inline src/app/core/models/report-unflattened.model.ts:10]: Can we clean up this mixed data type if possible? create another variable called approvalsMap if that makes it easier

File: src/app/core/services/report.service.spec.ts [hunk 4]
```diff
@@ -248,29 +262,56 @@ describe('ReportService', () => {
...
-    mockReports();
-    spyOn(dateService, 'fixDates').and.returnValues(
-      apiReportRes.data[0],
-      apiReportRes.data[1],
-      apiReportRes.data[2],
-      apiReportRes.data[3]
-    );
+  describe('getMyReports()', () => {
+    it('should get reports from API as specified by params', (done) => {
+      mockExtendedOrgUser();
```
Comment (Julias0) [inline src/app/core/services/report.service.spec.ts:267]: why have this inside a function - this is 1 line only so it doesn't really help with reuse or rewriting. Same for mockReports
The code in the line is more obvious than the method call
Ref: https://refactoring.guru/inline-method

File: src/app/core/services/report.service.spec.ts [hunk 5]
```diff
@@ -248,29 +254,56 @@ describe('ReportService', () => {
...
+          or: [],
+        },
+        order: 'rp_created_at.desc',
+      };
 
-    reportService.getMyReports(params).subscribe((res) => {
-      expect(res).toEqual(res);
-      expect(apiv2Service.get).toHaveBeenCalledTimes(1);
-      done();
+      reportService.getMyReports(params).subscribe((res) => {
```
Comment (Julias0) [inline src/app/core/services/report.service.spec.ts:277]: can we also check apiv2Service.get toHaveBeenCalledWith. That will test if the params are being properly added or not

File: src/app/core/services/report.service.spec.ts [hunk 6]
```diff
@@ -248,29 +254,56 @@ describe('ReportService', () => {
...
+        expect(res).toEqual(res);
+        expect(apiv2Service.get).toHaveBeenCalledTimes(1);
+        done();
+      });
+    });
+
+    it('should get reports from API when no order is specified', (done) => {
```
Comment (Julias0) [inline src/app/core/services/report.service.spec.ts:284]: Same as above - we should test the toHaveBeenCalledWith of apiv2Service.get, to see if the proper params are being sent or not.

File: src/app/core/models/api-reports.model.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+import { ExtendedReport } from './report.model';
+
+export interface ApiV2Reports {
```
Comment (rvab) [inline src/app/core/models/api-reports.model.ts:3]: There's already an `APIV2Response` model. Why do you need a new model for `ApiV2Reports`?

File: src/app/core/services/report.service.spec.ts [hunk 8]
```diff
@@ -605,18 +680,20 @@ describe('ReportService', () => {
       expect(res).toEqual(apiReportUpdatedDetails);
       expect(apiService.post).toHaveBeenCalledOnceWith('/reports', apiErptReporDataParam.rp);
       expect(reportService.clearTransactionCache).toHaveBeenCalledTimes(1);
-      expect(dataTransformService.unflatten).toHaveBeenCalledOnceWith(reportParam);
+      expect(dataTransformService.unflatten).toHaveBeenCalledWith(reportParam);
+      expect(dataTransformService.unflatten).toHaveBeenCalledTimes(1);
```
Comment (rvab) [inline src/app/core/services/report.service.spec.ts:684]: why remove `toHaveBeenCalledOnceWith`? isn't it short hand for `toHaveBeenCalledWith` and `toHaveBeenCalledTimes`?

File: src/app/core/services/report.service.spec.ts [hunk 9]
```diff
@@ -605,18 +680,20 @@ describe('ReportService', () => {
...
-    reportService.getReportPermissions(orgSettingsParams).subscribe((res) => {
-      expect(res).toEqual(reportAllowedActionsResponse);
-      expect(permissionsService.allowedActions).toHaveBeenCalledTimes(1);
-      done();
+  describe('getReportPermissions()', () => {
+    it('should get report permissions', (done) => {
+      permissionsService.allowedActions.and.returnValue(of(reportAllowedActionsResponse));
+      reportService.getReportPermissions(orgSettingsParams).subscribe((res) => {
+        expect(res).toEqual(reportAllowedActionsResponse);
+        expect(permissionsService.allowedActions).toHaveBeenCalledTimes(1);
```
Comment (rvab) [inline src/app/core/services/report.service.spec.ts:694]: `toHaveBeenCalledOnceWith` isn't required?

File: src/app/core/services/report.service.spec.ts [hunk 10]
```diff
@@ -913,30 +991,74 @@ describe('ReportService', () => {
...
+    spyOn(reportService, 'searchParamsGenerator').and.returnValue({
+      state: ['DRAFT', 'APPROVER_PENDING', 'APPROVER_INQUIRY'],
+    });
+    spyOn(reportService, 'getPaginatedERptcCount').and.returnValue(of({ count: 2 }));
+    spyOn(reportService, 'getPaginatedERptc').and.returnValue(of(addApproverERpts));
+    spyOn(reportService, 'getApproversInBulk').and.returnValue(of(addApproversParam));
+    spyOn(reportService, 'addApprovers').and.returnValue(expectedAddedApproverERpts);
+
+    reportService.getFilteredPendingReports({ state: 'edit' }).subscribe((res) => {
+      expect(res).toEqual(expectedAddedApproverERpts);
```
Comment (rvab) [inline src/app/core/services/report.service.spec.ts:1010]: You should be checking for `toHaveBeenCalledOnceWith` for all the mocked methods(line no 1004 to 1007)

**Review by rvab — CHANGES_REQUESTED** [review]

**Review by Dimple16 — CHANGES_REQUESTED** [review]
added comments

File: src/app/core/services/report.service.spec.ts [hunk 11]
```diff
@@ -248,29 +252,72 @@ describe('ReportService', () => {
...
-      order: 'rp_created_at.desc',
-    };
+      };
 
-    reportService.getMyReports(params).subscribe((res) => {
-      expect(res).toEqual(res);
-      expect(apiv2Service.get).toHaveBeenCalledTimes(1);
-      done();
+      reportService.getMyReports(params).subscribe((res) => {
+        expect(res).toEqual(res);
```
[the remark this answers is not in the document]
Comment (jayfyle) [reply src/app/core/services/report.service.spec.ts]: The function returns the data as given as response from the API after date manipulation, I have changed the statement.

File: src/app/core/services/report.service.spec.ts [hunk 12]
```diff
@@ -248,29 +252,72 @@ describe('ReportService', () => {
...
+      const apiParams = {
+        offset: 0,
+        limit: 10,
+        order: 'rp_created_at.desc,rp_id.desc',
+        rp_org_user_id: 'eq.ouX8dwsbLCLv',
+        or: [],
+      };
+
+      reportService.getMyReports(params).subscribe((res) => {
+        expect(res).toEqual(res);
```
[the remark this answers is not in the document]
Comment (jayfyle) [reply src/app/core/services/report.service.spec.ts]: Same as above

File: src/app/core/services/report.service.spec.ts [hunk 13]
```diff
@@ -651,50 +729,72 @@ describe('ReportService', () => {
...
+      const getMyReportsParam1 = {
+        offset: 0,
+        limit: 2,
+        queryParams: {
+          rp_state: 'in.(DRAFT,APPROVER_PENDING,APPROVER_INQUIRY)',
+        },
+        order: undefined,
+      };
+
+      const getMyReportsParam2 = {
```
[the remark this answers is not in the document]
Comment (jayfyle) [reply src/app/core/services/report.service.spec.ts]: There is no model for API parameters, I am creating a common model, that we can use for methods where we pass API params.

File: src/app/core/services/report.service.ts [hunk 14]
```diff
@@ -507,34 +507,7 @@ export class ReportService {
...
-        toDate = new Date(new Date(search.dateRange.to).setHours(23, 59, 59, 999));
-      }
-
-      dateParams = {
-        created_at: ['gte:' + new Date(fromDate).toISOString(), 'lte:' + new Date(toDate).toISOString()],
-      };
-    }
-
-    return Object.assign({}, params, searchParams, dateParams);
+    return Object.assign({}, params, searchParams);
```
[the remark this answers is not in the document]
Comment (jayfyle) [reply src/app/core/services/report.service.ts:518]: All the instances of the method being called are in the expense forms, where we don't pass `dateRange` property. Everywhere there is only state property being passed. But I will thoroughly test the relevant flows once the PR is merged

File: src/app/core/models/api-params.model.ts [hunk 15]
```diff
@@ -0,0 +1,10 @@
+export interface ApiParams {
```
Comment (Dimple16) [inline src/app/core/models/api-params.model.ts:1]: this looks very specific to report API params, rename this to `ReportApiParams`

**Review by Dimple16 — APPROVED** [review]

File: src/app/core/services/report.service.ts [hunk 16]
```diff
@@ -452,7 +457,9 @@ export class ReportService {
     return this.apiService.post('/reports/summary/download', data);
   }
 
-  getAllExtendedReports(config: Partial<{ order: string; queryParams: any }>): Observable<ExtendedReport[]> {
+  getAllExtendedReports(
+    config: Partial<{ order: string; queryParams: ReportApiParams['queryParams'] }>
```
Comment (rvab) [inline src/app/core/services/report.service.ts:461]: same as above

File: src/app/core/services/report.service.ts [hunk 17]
```diff
@@ -331,7 +331,12 @@ export class ReportService {
   }
 
   getMyReports(
-    config: Partial<{ offset: number; limit: number; order: string; queryParams: any }> = {
+    config: Partial<{
+      offset: number;
+      limit: number;
+      order: string;
+      queryParams: ReportApiParams['queryParams'];
```
Comment (rvab) [inline src/app/core/services/report.service.ts:338]: is it recommended to use `ReportApiParams['queryParams']`?

File: src/app/core/services/report.service.ts [hunk 18]
```diff
@@ -246,7 +247,6 @@ export class ReportService {
           if (res.data.next_at) {
```
Comment (rvab) [inline src/app/core/services/report.service.ts:247]: if you're checking for `res.data.next_at`, then it means `next_at` is optional. Why did you make it mandatory in the `ReportAutoSubmissionDetails` model?

**Review by rvab — CHANGES_REQUESTED** [review]

File: src/app/core/services/report.service.ts (hunk 18 above)
Comment (jayfyle) [reply src/app/core/services/report.service.ts:247]: Was trying out something to get more coverage, forgot to change it back. Have restored it to the previous version.

**Review by rvab — APPROVED** [review]

**Pull request merged**